zio-config-magnolia/src/main/scala/zio/config/magnolia/ConfigDescriptorProvider.scala

```diff
@@ -0,0 +1,101 @@
+package zio.config.magnolia
+
+import zio.config.{ ConfigDescriptor }
+import ConfigDescriptor._
+import magnolia._
+
+import scala.language.experimental.macros
+
+trait ConfigDescriptorProvider[T] {
+  def getDescription(path: Vector[String]): ConfigDescriptor[String, String, T]
+}
+
+object ConfigDescriptorProvider {
+  def apply[T](implicit ev: ConfigDescriptorProvider[T]): ConfigDescriptorProvider[T] = ev
+
+  def instance[T](f: String => ConfigDescriptor[String, String, T]): ConfigDescriptorProvider[T] =
+    new ConfigDescriptorProvider[T] {
+      override def getDescription(path: Vector[String]): ConfigDescriptor[String, String, T] = {
+        def loop(
+          tail: List[String],
+          acc: ConfigDescriptor[String, String, T]
+        ): ConfigDescriptor[String, String, T] =
+          tail match {
+            case Nil =>
+              acc
+
+            case h :: Nil =>
+              loop(Nil, f(h))
+
+            case h :: tail =>
+              nested(h)(loop(tail, acc))
+          }
+
+        loop(
+          path.toList,
+          empty[String, String, T].xmapEither[T]({
+            case Some(v) => Right(v)
+            case None    => Left("failed - undefined state")
+          })(v => Right(Some(v)))
+        )
+      }
+    }
+
+  implicit val stringDescriptor: ConfigDescriptorProvider[String] =
+    instance(string)
+
+  implicit val doubleDescriptor: ConfigDescriptorProvider[Double] =
+    instance(double)
+
+  implicit val intDescriptor: ConfigDescriptorProvider[Int] =
+    instance(int)
+
+  implicit def opt[A: ConfigDescriptorProvider]: ConfigDescriptorProvider[Option[A]] =
+    ConfigDescriptorProvider[A].getDescription(_).optional
+
```

Comment:
You probably planned to add it later, but things like `List` also need a custom instance otherwise it's going to derive it (`List` being a sealed abstract class) and give you weird results.

Reply:
Yes, I need to include the whole list.